[quant] Eager mode equalization support for ConvReLU and LinearReLU

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58792 [quant] Eager mode equalization support for ConvReLU and LinearReLU**

Summary: Enabling support for fused modules like ConvReLU or LinearReLU on eager mode cross-layer equalization.

Test Plan: `python test/test_quantization.py TestEqualizeEager`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28647242](https://our.internmc.facebook.com/intern/diff/D28647242)

angelayi committed:
[quant] Eager mode equalization support for ConvReLU and LinearReLU

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/quantization/_equalize.py:
```diff
@@ -131,6 +142,11 @@ def converged(curr_modules, prev_modules, threshold=1e-4):
     for name in curr_modules.keys():
-        difference = curr_modules[name].weight.sub(prev_modules[name].weight)
+        curr_weight = curr_modules[name].weight if type(
+            curr_modules[name]) in _supported_types else curr_modules[name][0].weight
```
We seem to be using the logic `do a if x in _supported_types else b`. Maybe we can move this logic out to a helper function and invoke it in all the call sites where relevant?

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -38,16 +39,18 @@ def cross_layer_equalization(module1, module2, output_axis=0, input_axis=1):
     ranges of the second tensors' input channel
     '''
-    if type(module1) not in _supported_types or type(module2) not in _supported_types:
+    if (type(module1) not in _supported_types or type(module2) not in _supported_types) and \
+       (type(module1) not in _supported_intrinsic_types or type(module2) not in _supported_intrinsic_types):
```
can you combine _supported_types and _supported_intrinsic_types in the same set and do the original check?
```
if type(module1) not in _all_supported_types or type(module2) not in _all_supported_types:
```
I think the current check will miss something

angelayi replied:
Could I create another global `_all_supported_types` which is the union of `_supported_types` and `_supported_intrinsic_types`? I wanted to keep `_supported_types` and `_supported_intrinsic_types` separate so that we could case on them when getting the weights. Would this be ok?

jerryzh168 commented on torch/quantization/_equalize.py:
```diff
@@ -72,9 +75,17 @@ def cross_layer_equalization(module1, module2, output_axis=0, input_axis=1):
-    module1.bias = torch.nn.Parameter(bias)
-    module2.weight = torch.nn.Parameter(weight2)
+    if type(module1) in _supported_types:
+        module1.weight = torch.nn.Parameter(weight1)
+        module1.bias = torch.nn.Parameter(bias)
```
these assignment can also be refactored to have a set_weight/set_bias function as well

angelayi committed:
Update on "[quant] Eager mode equalization support for ConvReLU and LinearReLU"


Summary: Enabling support for fused modules like ConvReLU or LinearReLU on eager mode cross-layer equalization.

Test Plan: `python test/test_quantization.py TestEqualizeEager`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28647242).

jerryzh168 reviewed (approved):
Looks good, thanks!
